### PR TITLE
Cleanup group code

### LINKED ISSUE #829: `create_group` probably should throw a custom exception
Right now there's a call to `error`:

https://github.com/JuliaIO/HDF5.jl/blob/ef228d57be54d2082a6a7581895fd4becbfe3714/src/HDF5.jl#L723

In my code, I'd like to overwrite the data in a group _if it already exists_, but currently that would make me catch a `ErrorException`:

```julia
try
    create_group(fid, grp_name)
catch e
   if e isa ErrorException
      @warn "Replacing data!"
   else
      throw(e)
   end
end
```

However, this doesn't tell me anything about where or why the `ErrorException` occurred: maybe the file was corrupt, maybe it got randomly deleted on disk - I can't tell.

It would be nice to have a _type_ of exception that would explain what caused it and let me decide how to handle it. Something like this:

```julia
try
    create_group(fid, grp_name)
catch e
   if e isa GroupExistsException
      @warn "Replacing data!"
   else
      throw(e)
   end
end
```

### PR DESCRIPTION
**Pull request by simonbyrne**
- Clarifies and adds docs
- Simplifies interface to `create_group`, avoiding mutation of default properties
- Closes properties after completion (helps #990)
- Removes custom exception for `create_group` (fixes #829)

### CODE REVIEW

File: src/groups.jl [hunk 1]
```diff
@@ -1,43 +1,43 @@
 """
-    create_group(parent, path[, lcpl, gcpl]; properties...)
+    create_group(parent::Union{File,Group}, path::AbstratString; properties...)
```
Comment (mkitti) [inline src/groups.jl:2]:
```suggestion
    create_group(parent::Union{File,Group}, path::AbstractString; properties...)
```

**Review by mkitti — CHANGES_REQUESTED** [review]
Could we separate the breaking and non-breaking components of this? 

For the breaking portion, we will probably want to use the `HDF5Context` structure to initialize the property lists rather than continuing to use the globals. We also should still support but deprecate the old methods.

**Comment by simonbyrne** [comment]
What parts do you consider breaking?

**Comment by simonbyrne** [comment]
I added a deprecation warning if they're both used.

**Comment by musm** [comment]
> For the breaking portion, we will probably want to use the HDF5Context structure to initialize the property lists rather than continuing to use the globals. We also should still support but deprecate the old methods.

Let's save the HDF5Context changes for a separate PR

**Review by musm — APPROVED** [review]

**Review by mkitti — APPROVED** [review]
With the `depwarn`, this works for me.

**Pull request merged**